### PR TITLE
Ldap support

### PR DESCRIPTION
**Pull request by arturo-seijas**

### CODE REVIEW

**Review by jdkandersson — DISMISSED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -609,7 +630,48 @@ def _get_indico_env_config(self, container: Container) -> Dict:
...
+                    "ubuntu_ldap": {
+                        "type": "ldap",
+                        "title": "LDAP",
+                        "ldap": _ldap_config,
+                        "mapping": {
+                            "email": "mail",
+                            "affiliation": "o",
+                            "first_name": "givenName",
+                            "last_name": "sn",
+                            "phone": "mobile",
```
Comment (mthaddon) [inline src/charm.py:660]: Are we really planning to sync phone numbers? Is that required by Indico somehow? Seems potentially problematic or unexpected.

File: src/charm.py [hunk 2]
```diff
@@ -609,7 +630,48 @@ def _get_indico_env_config(self, container: Container) -> Dict:
                     "identifier_field": "openid",
                 }
             }
+            if self.config["ldap_host"]:
+                _ldap_config = {
```
Comment (mthaddon) [inline src/charm.py:634]: It seems odd that we're pulling some variables here (`ldap_host`) and hard coding others (`cn=Indico Bot,ou=bots,dc=canonical,dc=com`). I realise we're only supporting one provider currently, but surely this should be checking if the host matches what we're expecting and mapping values to that? Maybe just `if self.config["ldap_host"] == "ldap.canonical.com"`?

File: .github/workflows/integration_test.yaml [hunk 3]
```diff
@@ -12,3 +12,4 @@ jobs:
       chaos-experiments: pod-delete
       chaos-app-label: app.kubernetes.io/name=indico
       load-test-enabled: true
+      load-test-run-args: "-e LOAD_TEST_HOST=localhost"
```
Comment (mthaddon) [inline .github/workflows/integration_test.yaml:15]: Is this intended, seems unrelated?

File: config.yaml [hunk 4]
```diff
@@ -38,6 +38,14 @@ options:
     type: string
     description: HTTPS proxy URL to be used when installing plugins or downloading customization files, e.g. 'http://squid.example:3128'.
     default: ''
+  ldap_host:
+    type: string
+    description: URL for the LDAP server to use as identity provider. Requires SAML to be configured.
+    default: ''
+  ldap_password:
+    type: string
+    description: LDAP for the LDAP server to use as identity provider.
```
Comment (mthaddon) [inline config.yaml:47]: This description doesn't quite make sense. Can you reword it?

File: config.yaml [hunk 5]
```diff
@@ -38,6 +38,14 @@ options:
     type: string
     description: HTTPS proxy URL to be used when installing plugins or downloading customization files, e.g. 'http://squid.example:3128'.
     default: ''
+  ldap_host:
+    type: string
+    description: URL for the LDAP server to use as identity provider. Requires SAML to be configured.
```
Comment (gregory-schiano) [inline config.yaml:43]: Is it and URI or a host ?
If that's an URI I'd call it `ldap_uri`

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:634]: Not sure I follow. I see only two variable in that block: ldap_host and ldap_password

File: config.yaml (hunk 4 above)
Comment (gregory-schiano) [reply config.yaml]: Yeap he probably meant `Password for the LDAP server`

Comment (arturo-seijas) [reply config.yaml]: Done

File: .github/workflows/integration_test.yaml (hunk 3 above)
Comment (arturo-seijas) [reply .github/workflows/integration_test.yaml:15]: It is, we have some checks failing because of that, though we will probably remove this check on PR submission

File: indico.Dockerfile [hunk 6]
```diff
@@ -20,14 +26,17 @@ ENV DEBIAN_FRONTEND=noninteractive \
...
-
-RUN addgroup --gid ${indico_gid} indico \
-    && adduser --system --gid ${indico_gid} --uid ${indico_uid} --home /srv/indico indico
+RUN apt-get update \
+    && apt-get install -y gettext git libxmlsec1-dev locales postgresql-client python3-pip texlive-xetex \
+    && rm -rf /var/lib/apt/lists/* \
+    && addgroup --gid ${indico_gid} indico \
+    && adduser --system --gid ${indico_gid} --uid ${indico_uid} --home /srv/indico indico \
+    && /bin/bash -c "mkdir -p --mode=775 /srv/indico/{archive,cache,custom,etc,log,tmp}" \
+    && chown indico:indico /srv/indico  /srv/indico/cache /srv/indico/archive /srv/indico/custom \
```
Comment (gregory-schiano) [inline indico.Dockerfile:35]: Why not using the {} as in the previous command here ?

File: src/charm.py [hunk 7]
```diff
@@ -478,22 +479,56 @@ def _get_redis_cache_rel(self) -> Optional[Relation]:
...
+        Returns:
+            List containing the installed plugins.
+        """
+        process = container.exec(
+            ["/srv/indico/.local/bin/indico", "setup", "list-plugins"],
+            user="indico",
+        )
+        output, _ = process.wait_output()
+        # Parse output table, discarding header and footer rows and fetching first column value
+        return [item.split("|")[1].strip() for item in output.split("\n")[3:-2]]
```
Comment (gregory-schiano) [inline src/charm.py:530]: Did you test with no plugins installed ?

P.S.: Ah I see this was pre-existing code, but could worth it checking this edge case

File: indico.Dockerfile [hunk 8]
```diff
@@ -1,16 +1,22 @@
...
+    && apt-get install -y libldap2-dev libpq-dev libsasl2-dev libssl-dev libxmlsec1-dev pkg-config python3-pip \
+    && rm -rf /var/lib/apt/lists/* \
+    && addgroup --gid ${indico_gid} indico \
+    && adduser --system --gid ${indico_gid} --uid ${indico_uid} --home /srv/indico indico
 
 ENV UWSGI_EMBED_PLUGINS=stats_pusher_statsd
-RUN pip install --prefer-binary indico==3.2 indico-plugin-piwik indico-plugin-storage-s3 python3-saml uwsgi
+
+USER indico
+RUN python3 -m pip install --no-cache-dir --prefer-binary indico==3.2 indico-plugin-piwik indico-plugin-storage-s3 python-ldap python3-saml uwsgi
```
Comment (nrobinaubertin) [inline indico.Dockerfile:15]: I'm not completely sure, but I thought that to install to `$HOME/.local`, you needed to add `--user` to the `pip install` command

File: src/charm.py (hunk 2 above)
Comment (gregory-schiano) [reply src/charm.py:634]: This charm is not "Canonical" only, so indeed these variables shouldn't be hardcoded as it'll make it unusable for other users.
Usually on application providing LDAP connection provides at least the following configuration params:
- bind DN
- user DN
- user filter
- group DN
- group filter
- And fields to fetch for username, email, id, group_name, ...

File: src/charm.py [hunk 9]
```diff
@@ -609,7 +630,48 @@ def _get_indico_env_config(self, container: Container) -> Dict:
...
+                    "user_base": "ou=staff,dc=canonical,dc=com",
+                    "user_filter": "(objectClass=canonicalPerson)",
+                    "gid": "cn",
+                    "group_base": "dc=canonical,dc=com",
+                    "group_filter": "(objectClass=canonicalTeam)",
+                    "member_of_attr": "mozillaCustom1",
+                    "ad_group_style": False,
+                }
+                identity_providers = {
+                    "ubuntu_ldap": {
```
Comment (gregory-schiano) [inline src/charm.py:651]: I'd make this naming generic like "ldap_provider"

File: src/charm.py [hunk 10]
```diff
@@ -609,7 +630,48 @@ def _get_indico_env_config(self, container: Container) -> Dict:
...
+                            "first_name": "givenName",
+                            "last_name": "sn",
+                            "phone": "mobile",
+                        },
+                        "trusted_email": True,
+                        "synced_fields": {"first_name", "last_name", "affiliation", "phone"},
+                    }
+                }
+                provider_map = {
+                    "ubuntu": {
```
Comment (gregory-schiano) [inline src/charm.py:667]: I'd make this name generic like `provider`

File: src/charm.py [hunk 11]
```diff
@@ -647,6 +713,15 @@ def _on_config_changed(self, event: HookEvent) -> None:
             )
             event.defer()
             return
+        if not self._is_ldap_host_valid():
+            self.unit.status = BlockedStatus(
+                f"Invalid ldap_host option provided. Only {CANONICAL_LDAP_HOST} is available."
+            )
+            event.defer()
+            return
```
Comment (gregory-schiano) [inline src/charm.py:721]: Why limiting to Canonical LDAP only ?

File: src/charm.py [hunk 12]
```diff
@@ -478,22 +479,56 @@ def _get_redis_cache_rel(self) -> Optional[Relation]:
...
         broker_port = ""
-        if (broker_rel := self._get_redis_broker_rel()) is not None:
-            broker_unit = next(
-                unit for unit in broker_rel.data if unit.name.startswith("redis-broker")
-            )
+        if (broker_rel := self._get_redis_rel(name)) is not None:
+            broker_unit = next(unit for unit in broker_rel.data if unit.name.startswith(name))
             broker_host = broker_rel.data[broker_unit].get("hostname")
             broker_port = broker_rel.data[broker_unit].get("port")
         return f"redis://{broker_host}:{broker_port}"
```
Comment (nrobinaubertin) [inline src/charm.py:497]: Since this function is for redis cache and broker, I would avoid using the words broker and cache (and replace them by just "redis")

**Review by nrobinaubertin — CHANGES_REQUESTED** [review]

File: config.yaml (hunk 5 above)
Comment (arturo-seijas) [reply config.yaml:43]: Host, not URI `"uri": f"ldaps://{self.config['ldap_host']}"`

File: src/charm.py (hunk 11 above)
Comment (arturo-seijas) [reply src/charm.py:721]: It would be quite complex to pass all these configs, so that's the scope for now, same as for SAML

File: src/charm.py (hunk 9 above)
Comment (arturo-seijas) [reply src/charm.py:651]: It won't be reusable anyway

File: src/charm.py (hunk 10 above)
Comment (arturo-seijas) [reply src/charm.py:667]: It won't be reusable anyway

File: indico.Dockerfile (hunk 8 above)
Comment (arturo-seijas) [reply indico.Dockerfile:15]: Running as the user also does the trick

File: src/charm.py (hunk 12 above)
Comment (arturo-seijas) [reply src/charm.py]: Indeed. Fixed

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:634]: It would be quite complex to pass all these configs, so that's the scope for now, same as for SAML

File: src/charm.py (hunk 1 above)
Comment (mthaddon) [reply src/charm.py:660]: Yes please, let's remove it.

Comment (arturo-seijas) [reply src/charm.py]: Done

File: indico.Dockerfile (hunk 6 above)
Comment (arturo-seijas) [reply indico.Dockerfile]: Indeed. Changed

File: src/charm.py (hunk 2 above)
Comment (mthaddon) [reply src/charm.py:634]: I agree we should try to avoid having to specify all those values in juju config if possible, and I think the approach of only requiring ldap_host and ldap_password makes sense, but my comment was about the fact that we're checking `if self.config["ldap_host"] but then defining `_ldap_config` to include some hard coded and some parameterised values. If we add support for other LDAP providers in the future (or if someone like me is reading through the code) it might be a bit confusing.

**Review by nrobinaubertin — DISMISSED** [review]
LGTM

File: src/charm.py (hunk 9 above)
Comment (arturo-seijas) [reply src/charm.py]: Done

File: src/charm.py (hunk 10 above)
Comment (arturo-seijas) [reply src/charm.py]: Done

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:634]: Changed

File: src/charm.py [hunk 13]
```diff
@@ -35,6 +35,7 @@
 NGINX_PROMEXP_PORT = "9113"
 STATSD_PROMEXP_PORT = "9102"
 UBUNTU_SAML_URL = "https://login.ubuntu.com/saml/"
+CANONICAL_LDAP_HOST = "ldap.canonical.com"
```
Comment (amandahla) [inline src/charm.py:38]: alphabetical order?

File: src/charm.py [hunk 14]
```diff
@@ -626,6 +687,10 @@ def _get_http_proxy_configuration(self) -> Dict[str, str]:
             config["HTTPS_PROXY"] = self.config["https_proxy"]
         return config
 
+    def _is_ldap_host_valid(self) -> bool:
+        """Check if the LDAP hostis currently supported."""
```
Comment (amandahla) [inline src/charm.py:691]: Minor thing: host is

File: src/charm.py (hunk 13 above)
Comment (arturo-seijas) [reply src/charm.py]: Fixed

File: src/charm.py (hunk 14 above)
Comment (arturo-seijas) [reply src/charm.py]: Fixed. Thanks for your input

**Review by jdkandersson — APPROVED** [review]

File: src/charm.py (hunk 11 above)
Comment (gregnebu) [reply src/charm.py:720]: As discussed since my comment, we'll keep the implementation as it is (which means limiting to canonical LDAP only) but will work on integrator charms for LDAP / SAML and will refactor this part accordingly

Comment (gregory-schiano) [reply src/charm.py:720]: As discussed since my comment, we'll keep the implementation as it is (which means limiting to canonical LDAP only) but will work on integrator charms for LDAP / SAML and will refactor this part accordingly

**Review by gregory-schiano — APPROVED** [review]
LGTM
